docs/en/02_Developer_Guides/15_Customising_the_Admin_Interface/How_Tos/Extend_CMS_Interface.md

```diff
@@ -204,26 +204,443 @@ To make the actions more user-friendly you can also use alternating buttons as
+ [React Dev Tools](https://chrome.google.com/webstore/detail/react-developer-tools/fmkadmapgofadopljbjfkapdkoienihi?hl=en).
+ 
+ For this, you can use the third parameter of the `update()` function. It takes an arbitrary
+ name for the enhancement you're applying.
+ 
+ __module-a/js/main.js__
+ ```js
+ (update) => update('TextField', CharacterCounter, 'CharacterCounter')
+ ```
+ __module-b/js/main.js__
+ ```js
+ (update) => update('TextField', TextLengthChecker, 'TextLengthChecker')
+ ```
+### Dealing with events
+Let's make a new customisation that customises the behaviour of a button. We'll have
+all form actions throw a `window.confirm()` message before executing their action. Further,
+we'll apply some new style to the button if it is in a loading state.
+
+__my-module/js/components/ConfirmingFormButton.js__
+```js
+import React from 'react';
+
+export default (FormAction) => (props) => {
+  const newProps = {
+    ...props,
+    data: {
+      ...props.data,
+      buttonStyle: props.loading ? 'danger' : props.data.buttonStyle
+    },
+    handleClick(e) {
+      if(window.confirm('Did you really mean to click this?')) {
+        props.handleClick(e);
+      }
+    }
+  }
+
+  return <FormAction {...newProps} />
+}
+```
+
+__my-module/js/main.js__
+```js
+import ConfirmingFormButton from './components/ConfirmingFormButton';
+
+Injector.transform('my-transformation', (update) => {
+  update('FormAction', ConfirmingFormButton, 'ConfirmingFormButton');
+});
+```
+### Registering new React components
+
+If you've created a module using React, it's a good idea to afford other developers an 
+API to enhance those components. To do that, simply register them with `Injector`.
+
+__my-public-module/js/main.js__
+```js
+import Injector from 'lib/Injector';
+
+Injector.register('MyComponent', MyComponent);
+```
+
+Now other developers can customise your components with `Injector.update()`.
+
+Note: Overwriting components by calling `register()` multiple times for the same
+service name is discouraged, and will throw an error. Should you really need to do this,
+you can pass `{ force: true }` as the third argument to the `register()` function.
+
+### Using the injector directly within your component
+
+On rare occasions, you may just want direct access to the injector in your component. If
+your dependency requirements are dynamic, for example, you won't be able to explicitly
+declare them in `inject()`. In cases like this, use `withInjector()`. This higher order
+component puts the `Injector` instance in `context`.
+
+```js
+class MyGallery extends React.Component {
+  render () {
+    <div>
+      {this.props.items.map(item => {
+        const Component = this.context.injector.get(item.type);
+        return <Component title={item.title} image={item.image} />
+      })}
+    </div>
+  }
+}
+
+export default withInjector(MyGallery);
+```
 
 ### Interfacing with legacy CMS JavaScript
```

Comment:
Likely not, but getting into that type of cleanup is out of scope for this story. I think it's fair to say that this section of the documentation needs a thorough review.